components/brave_shields/browser/brave_shields_web_contents_observer.cc

```diff
@@ -250,4 +261,58 @@ void BraveShieldsWebContentsObserver::RegisterProfilePrefs(
   registry->RegisterUint64Pref(kFingerprintingBlocked, 0);
 }
 
+void BraveShieldsWebContentsObserver::ReadyToCommitNavigation(
+    content::NavigationHandle* navigation_handle) {
+  auto frame_tree_node_id = navigation_handle->GetFrameTreeNodeId();
+  auto *frame_tree_node = content::FrameTreeNode::GloballyFindByID(
+      frame_tree_node_id);
+  auto* navigation_entry =
+    frame_tree_node->navigator()->GetController()->GetPendingEntry();
+
+  if (!navigation_entry) {
+    navigation_entry =
+      frame_tree_node->navigator()->GetController()->GetLastCommittedEntry();
+  }
+  GURL target_origin = navigation_handle->GetURL().GetOrigin();
+  if (!navigation_entry) {
+    return;
+  }
+  Profile* profile =
+      Profile::FromBrowserContext(web_contents()->GetBrowserContext());
+  GURL tab_origin(navigation_entry->GetURL().GetOrigin());
+
+  std::unique_ptr<base::Value> referrer_value =
+      HostContentSettingsMapFactory::GetForProfile(profile)
+      ->GetWebsiteSetting(
+          tab_origin, tab_origin, CONTENT_SETTINGS_TYPE_PLUGINS,
+          brave_shields::kReferrers, NULL);
+  ContentSetting referrer_setting =
+      content_settings::ValueToContentSetting(referrer_value.get());
+
+  std::unique_ptr<base::Value> shields_value =
+      HostContentSettingsMapFactory::GetForProfile(profile)
+      ->GetWebsiteSetting(
+          tab_origin, GURL(), CONTENT_SETTINGS_TYPE_PLUGINS,
+          brave_shields::kBraveShields, NULL);
+  ContentSetting shields_setting =
+      content_settings::ValueToContentSetting(shields_value.get());
+
+  auto original_referrer = navigation_handle->GetReferrer();
+  GURL site_instance(navigation_handle->GetStartingSiteInstance()->GetSiteURL());
+  // This does a TLD+1 comparison
+  bool sameTLDP1 = SameDomainOrHost(original_referrer.url,
+      target_origin, INCLUDE_PRIVATE_REGISTRIES);
+
+  if (referrer_setting != CONTENT_SETTING_ALLOW &&
+      shields_setting != CONTENT_SETTING_BLOCK &&
+      !sameTLDP1 &&
+      !brave::IsWhitelistedReferrer(tab_origin, target_origin) &&
+      !original_referrer.url.is_empty()) {
+    auto referrer = original_referrer;
+    referrer.url = navigation_handle->GetURL().GetOrigin();
+    navigation_entry->SetReferrer(
```

Comment:
this is an example of something that is different between the network delegate and the observer. I think this one is correct